### ISSUE TITLE
MoveToLocation causes main thead hangings and disconnections

### ISSUE THREAD

**Issue by theonlygusti** [MCCTeam/Minecraft-Console-Client#219]
Every time I use this to go on a server, it is super slow and will disconnect after a few moments.

`terrainandmovements` seems to be the cause. See [this pastebin](http://pastebin.com/wbhwQhHN) of my `MinecraftClient.ini` file.

**Comment by ORelio** [comment]
Hello,

I'm sorry but this is too vague for a bug report.

Please at least describe your configuration:
 - Your operating system with version
 - Your Mono version (if running Mac or Linux)
 - Minecraft version running on your server

**Comment by theonlygusti** [comment]
![http://i.cubeupload.com/u8qkuN.png](http://i.cubeupload.com/u8qkuN.png)

```
Mono JIT compiler version 4.4.2 (mono-4.4.0-branch-c7sr1/f72fe45 Wed Jul 27 16:20:13 EDT 2016)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  x86
	Disabled:      none
	Misc:          softdebug 
	LLVM:          yes(3.6.0svn-mono-master/a173357)
	GC:            sgen
```

Servers I've tried to use it with: Vanilla 1.10, CraftBukkit 1.10, Spigot 1.10, and (although I shouldn't really) mineplex.

**Comment by theonlygusti** [comment]
My `MinecraftClient.ini` is [in this pastebin](http://pastebin.com/CMzPgDqR), I am using what I think is the latest version:

    Console Client for MC 1.4.6 to 1.10.2 - v1.10.0 DEV - By ORelio & Contributors

**Comment by Pokechu22** [comment]
Try turning off `terrainandmovements` - it causes additional work to be done which may impact performance.  If it's still laggy with that disabled, then something else is clearly wrong.

Also... your login info is in that config - you might want to change your PW :/

**Comment by theonlygusti** [comment]
I need terrainandmovements

and yeah just realised I derped big time, just changed my password.

**Comment by ORelio** [comment]
Indeed, please change your password immediately (and edit/delete/replace your paste if you can), to remove your credentials so that your email also does not get spammed.

Regarding the report, thanks for the details!
The platform seems good, and you are indeed using a recent build. Nice!
The only thing that I'm not sure about is Mono 4.4, MCC has been successfully tested on Mono 4.0-4.2, that should be fine too with 4.4 but I can't confirm that.

In any case, is there some specific message while disconnecting, or is it fully silent?
Does disabling `terrainandmovements` as suggested by @Pokechu22 helps? (even if you need it)

**Comment by theonlygusti** [comment]
When it disconnects, it will say

     Connection has been lost

along with some weird, what looks like, an escape code. If you hang with a second I can get full output.

I have only started having this problem recently, and I don't know why it would be because I haven't changed anything since before the problem started.

**Comment by theonlygusti** [comment]
It sometimes, but not always, outputs a weird string like this: `>>747;3R` along with the `Connection has been lost` message.

**Comment by theonlygusti** [comment]
And I have just disabled terrainandmovements, it seems to be working better. Maybe it will quit in a bit, but since my last post 15 minutes ago it hasn't crashed, which is the longest it's run in a long time.

**Comment by ORelio** [comment]
Good! That means we got a hint about where the issue is located. Just in case, please re-enable terrain and movements, and use the latest [developpent build](https://ci.appveyor.com/project/ORelio/minecraft-console-client/build/artifacts), there is a slight change for issue #217 that _might_ help (but will proably not help, well, just in case...). If that doesn't help, I'll try to reproduce & debug your issue by using the same environment (well, at least, the same Mono version, MacOS is probably not the culprit)

**Comment by theonlygusti** [comment]
Ok thank you.

Now, not relating to what you just posted, but from before:

> And I have just disabled terrainandmovements, it seems to be working better. Maybe it will quit in a bit, but since my last post 15 minutes ago it hasn't crashed, which is the longest it's run in a long time.

Well, it did just crash. This was with the same configuration as when the issue was created but with terrainandmovements disabled.

```
Unhandled Exception:
OutOfMemoryException
[ERROR] FATAL UNHANDLED EXCEPTION: System.OutOfMemoryException: Out of memory
  at (wrapper managed-to-native) System.Object:__icall_wrapper_mono_gc_alloc_vector (intptr,intptr,intptr)
  at (wrapper alloc) System.Object:AllocVector (intptr,intptr)
  at System.Collections.Generic.Dictionary`2[TKey,TValue].Resize (Int32 newSize, Boolean forceNewHashCodes) <0x450be28 + 0x00067> in <filename unknown>:0 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].Resize () <0x450bde8 + 0x0002b> in <filename unknown>:0 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].Insert (System.Collections.Generic.TKey key, System.Collections.Generic.TValue value, Boolean add) <0x450b180 + 0x0025f> in <filename unknown>:0 
  at System.Collections.Generic.Dictionary`2[TKey,TValue].set_Item (System.Collections.Generic.TKey key, System.Collections.Generic.TValue value) <0x450b100 + 0x0006f> in <filename unknown>:0 
  at MinecraftClient.Mapping.Movement+<>c__DisplayClass3_0.<CalculatePath>b__0 () <0x4505ba8 + 0x0114f> in <filename unknown>:0 
  at System.Threading.ThreadHelper.ThreadStart_Context (System.Object state) <0x196ceb0 + 0x0008e> in <filename unknown>:0 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, Boolean preserveSyncCtx) <0x196b330 + 0x001bb> in <filename unknown>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, Boolean preserveSyncCtx) <0x196b2f0 + 0x00033> in <filename unknown>:0 
  at System.Threading.ExecutionContext.Run (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) <0x196b240 + 0x0005a> in <filename unknown>:0 
  at System.Threading.ThreadHelper.ThreadStart () <0x196d030 + 0x00039> in <filename unknown>:0 
```

**Comment by theonlygusti** [comment]
I am now trying with [the latest dev build](https://ci.appveyor.com/project/ORelio/minecraft-console-client/build/artifacts) as linked to by you.

**Comment by theonlygusti** [comment]
And it fails again. Exactly the same as with my previous build.

**Comment by ORelio** [comment]
Seems like there is also an "out of sync" issue even without terrain and movements. Oh, well...

**Comment by theonlygusti** [comment]
Enabling terrainandmovements in the `.ini` file but removing all my `MoveToLocation` calls from my C# scripts also doesn't have a problem.

**Comment by Pokechu22** [comment]
As in, it stops crashing?  Hm, that would indicate that (one of) your scripts is causing the problem - that, combined with the `OutOfMemoryException`, makes me think there's a memory leak somewhere (either in one of your scripts, or the code the script calls).

**Comment by theonlygusti** [comment]
I am only running one script, I am trying to debug it atm, but have not found any problems.

**Comment by ORelio** [comment]
MoveToLocation will trigger a pathfinding routine that is quite time consuming and is currently launched on the main thread. If it takes too much time, the server could silently disconnect the client, there is a 30-seconds timeout server-side. However that sould not happen normally because the pathfinding routine itselfs times out after 5 seconds. Unless... you are calling it 6 times in a row? 😅

**Comment by theonlygusti** [comment]
This is fine. It looks like it was something within my script.

I find it weird though that your pathfinding would cause the main thread to hang.

**Comment by ORelio** [comment]
Pathfinding is implemented with [A* Algorithm](https://en.wikipedia.org/wiki/A*_search_algorithm), which is very efficient to compute paths, but will handle unreachable locations pretty badly. In fact, if the requested path is impossible to find eg if you request to walk in the air or in a wall (which is not possible), the algorithm will try to circumvent an impossible to circumvent obstacle resulting in processing ALL the pathways in all the loaded chunks. That is why in MCC there is a 5-seconds timeout before stating the path cannot be found. And that is why there is an issue if you request pathfinding computation for an unreachable location more than 5 times in a row.

However if the path is possible and pretty straightforward, A* is optimized for that and will find the requested path pretty fast (usually less than a second in MCC for, let's say, less than 100 blocks). Here are great animations from Wikipedia: [Dijkstra](https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm#Algorithm) and [A*](https://en.wikipedia.org/wiki/A*_search_algorithm#Example). As you can see, A* is more efficient as it will try to go towards the goal, instead of processing everything first in order to find the absolutely most efficient path. That's waaay faster and almost optimal.

In MCC, for resource-sharing reasons, scripts are scheduled on the main thread, because some resources cannot be accessed in a consistent way otherwise. There is a similar issue for Bukkit plugins that run on the main Minecraft thread and can hang the server if badly coded. Of course as in Bukkit plugins, non-api related processings can (and should if they can be slow) be offloaded in separate threads (ie processings that does not use any Bukkit API calls or access Minecraft data).

Of course, pathfinding could be offloaded, but that would require careful coding. Example: The algorithm starts to compute pathfinding. A chunk gets unloaded on the main thread. It still tries to access that data. Ka-Boom. Let's admit we fix that. A chunk gets updated, causing other issues. And so on. This is not impossible however, if I recall well, Mojang offloaded it for hostile mobs (but I currently don't have that much time for optimizing that).

Hope this is more clear with these explanations :wink:

**Comment by theonlygusti** [comment]
Can we customise the time out? I would prefer a significantly shorter timeout for the pathfinding.

**Comment by ORelio** [comment]
Currently, no. This could be added if needed, however. Or, you can use [AutoTimeout](https://github.com/ORelio/Minecraft-Console-Client/blob/master/MinecraftClient/AutoTimeout.cs) for that:
````C#
AutoTimeout.Perform(() =>
{
    //Put code chunk that must timeout here
    MoveToLocation(/* ... */);
}, TimeSpan.FromSeconds(1) /* or directly milliseconds as an integer */);
````
This is an utility class already included in the Minecraft Console Client project.
In fact, the timeout in pathfinding is already implemented as described above. With 5s timeout.

**Comment by milutinke** [comment]
This has been fixed recently with the #2124.
